### ISSUE TITLE
Opportunistic compression?

### ISSUE THREAD

**Issue by lechner** [rfjakob/gocryptfs#257]
Hi Jakob,

Have you thought about offering opportunistic compression? In a networked setup, it may reduce traffic. Some file systems offer transparent compression at the disk level (for example, `btrfs`, which I use), but a fast client may be better equipped for the math. Either way, compression at a higher level may reduce the size of off-size backups and could make it safer to encrypt large amounts of data. Maybe I should look at other FUSE-based layers for compression, but I really like your software and thought I should ask. Thank you!

Best regards,
Felix

**Comment by rfjakob** [comment]
> Maybe I should look at other FUSE-based layers for compression, but I really like your software and thought I should ask.

:)

Encryption plus compression may be hard to implement *securely*, at least that's what I remember from the CRIME attack against SSL.

Then I am not sure how hard it is to get transparent compression implemented *correctly*. The most promising project I found is the 0.9 branch of `fusecompress`. It is maintained by Ulrich Hecht at https://code.google.com/archive/p/fusecompress/ , and the README looks good:

> This is the legacy C implementation of FuseCompress (0.9.x tree). It performs slightly less well than the current one implemented in C++ (1.99.x tree), but does not exhibit flaws such as data corruption or files growing larger and larger that make the current code (as of July 17 2008) unusable in a production environment.

> It works well enough that I have trusted it to handle the vast majority of my private data. I am also using it for my $HOME directory, source trees, and I have (with some really evil hacks) succeeded in installing an openSUSE 11.0 system with FuseCompress as its root filesystem.

The 2.x branch of fusecompress is maintained by Milan Svoboda at https://github.com/tex/fusecompress  , but the README is pretty scary:

>	Do NOT compile FuseCompress with g++-4.3.3! Strange threading problems
>	(broken locking) were detected (thank you Steve!) when using this version of
>	g++.

>	There is a memory leak (not too big, user detects it after milions of
>	copyied files) in 'file' package, more exactly in it's magic library, see:

**Comment by rfjakob** [comment]
(inadvertedly closed the ticket)

**Comment by rfjakob** [comment]
The only way I see to implement this is to compress 4KiB blocks independently, and have file holes between the compressed blocks. A compressed+encrypted file would look like this:
```
Compressed+encryted data blob @ offset 0
File hole (size depends on compressibility of data blob)
Compressed+encryted data blob @ offset 4KiB
File hole (size depends on compressibility of data blob)
...etc..
```

This needs FALLOC_FL_PUNCH_HOLE support from the backing filesystem. Luckily, the most common Linux filesystems seem to have it:
```
       Not all filesystems support FALLOC_FL_PUNCH_HOLE; if a filesystem
       doesn't support the operation, an error is returned.  The operation
       is supported on at least the following filesystems:
       *  XFS (since Linux 2.6.38)
       *  ext4 (since Linux 3.0)
       *  Btrfs (since Linux 3.7)
       *  tmpfs(5) (since Linux 3.5)
```
However, I'm not sure if this can work on NFS, and it will probably not work via Samba, so I'm not sure how useful it is.

Also, if you copy the encrypted files with a program that does not understand file holes, the holes are filled with actual zeros, and you lose any space savings.

**Comment by lechner** [comment]
As a radical suggestion, would it be possible to store each 4KiB block in a separate, underlying file? Thank you for looking into this!

**Comment by rfjakob** [comment]
Yes that would be possible. The downside is that renaming and deleting files is messy.

Still, CryFS seems to do just that: From https://www.cryfs.org/howitworks :

![storingdata](https://user-images.githubusercontent.com/286847/44309045-985bc700-a3c0-11e8-8da0-4d7e44bef7c9.png)

**Comment by lechner** [comment]
Also, on second thought many file systems may allocate physical blocks larger than the compressed content, which would negate any hoped-for space advantage.

**Comment by rfjakob** [comment]
Given that CryFS must already have the infrastructure for handling those fragments, I imagine adding compression would be relatively easy. Looks like there's a feature request already: https://github.com/cryfs/cryfs/issues/134

**Comment by brainchild0** [comment]
I discovered this issue upon investigating the possibility of better utilization of local disk, in contrast to the original intent of  optimizing network latency.

It would be very useful to have  compression support, especially if it can be selected at a granularity lower than an entire GFS store. 

The use of sparse file support in common Linux filesystems seems like a sound approach.   

True, it would not work on NFS, but would still be useful to many in the common use case of local disk, and implementation would be less complex than the case of a sequence of physical files for each logical file exposed via GFS. 

One concern is that compressed blocks may need to be large in order for padded regions  to benefit from optimization in the underlying store.  If a 4K block size is in use for encryption, then it may be necessary to determine a superblock granularity at which compression is occurring, making it considerably slower  for random-access I/O.  If this issue can be solved, then the idea may be worth considering.

Further, as observed earlier, the sparse representation is not in general preserved when files are copied between physical stores, but rather must be created by appropriate calls from  the application.

**Comment by Slartibartfast27** [comment]
I have major problems with RSS usage using the ubuntu 20.04 LTS version of armhf:
`` gocryptfs 1.7.1; go-fuse 0.0~git20190214.58dcd77; 2019-12-26 go1.13.5 linux/arm`` 
see https://github.com/rfjakob/gocryptfs/issues/569

**Comment by teras** [comment]
I found this thread having the hope that compression, if required, would be provided. But it seems it is not, right? Is this idea dead or there is still hope?
